haystack/nodes/base.py

```diff
@@ -74,6 +70,38 @@ def __init_subclass__(cls, **kwargs):
         # Enables generic load() for all specific component implementations.
         cls._subclasses[cls.__name__] = cls
 
+    @property
+    def name(self) -> Optional[str]:
+        if "name" not in self._component_config:
+            return None
+        return self._component_config["name"]
+
+    @name.setter
+    def name(self, value: str):
+        self._component_config["name"] = value
+
+    @property
+    def sub_components(self) -> List[BaseComponent]:
+        if "params" not in self._component_config:
+            return list()
+        return [param for param in self._component_config["params"].values() if isinstance(param, BaseComponent)]
+
+    @property
+    def type(self) -> str:
+        return self._component_config["type"]
+
+    def get_params(self, return_defaults: bool = False) -> Dict[str, Any]:
+        component_signature = self._get_signature()
+        params: Dict[str, Any] = {}
+        for key, value in self._component_config["params"].items():
+            if value != component_signature[key].default or return_defaults:
+                params[key] = value
+        if return_defaults:
+            for key, param in component_signature.items():
+                if key not in params:
+                    params[key] = param.default
+        return params
```

Comment:
Seems strangely convoluted. I think the following would work the same:
```python
def get_params(self, return_defaults: bool = False) -> Dict[str, Any]:
    explicit_params = deepcopy(self._component_config["params"])

    if not return_defaults:
        return explicit_params

    default_params = {key: param.default for key, param in self._get_signature().items()}
    return {**default_params, **explicit_params}
```
(untested)

Reply:
This version would not account for default values within explicit_params. I think we want to suppress them too if `return_defaults=False`. But I can definitely get rid of the second loop.

Reply:
I see what you mean, but on the other hand I would not necessarily expect this method to filter out defaults values if they were explicitly given. Keeping them would also make sure that if I load a YAML and I save it back, it's going to come out exactly as I loaded it, no differences, regardless of whether I gave the default parameters or not.

Not a big deal though, feel free to keep the current behavior :+1: